pkg/segment/aggregations/segaggs.go

```diff
@@ -2408,6 +2414,665 @@ func performValueColRequest(nodeResult *structs.NodeResult, aggs *structs.QueryA
+	if minSpan == nil {
+		return 0
+	}
+
+	switch minSpan.TimeScale {
+	case segutils.TMSecond:
+		return minSpan.Num
+	case segutils.TMMinute:
+		return minSpan.Num * 60
+	case segutils.TMHour:
+		return minSpan.Num * 3600
+	case segutils.TMDay:
+		return minSpan.Num * 86400
+	default:
+		// default returning num*(seconds in a month)
+		return minSpan.Num * 2592000
+	}
+}
+
+// These time ranges are estimated based on different queries executed in splunk, no documentation is present
+func findEstimatedTimeSpan(minValue float64, maxValue float64, maxBins uint64, minSpan *structs.BinSpanLength) *structs.BinSpanOptions {
+	minSpanSecs := getSecsFromMinSpan(minSpan)
+	intervalSec := (maxValue/1000 - minValue/1000) / float64(maxBins)
+	if minSpanSecs > intervalSec {
+		intervalSec = minSpanSecs
+	}
+	var num float64
+	timeUnit := segutils.TMSecond
+	if intervalSec < 1 {
+		num = 1
+	} else if intervalSec <= 10 {
+		num = 10
+	} else if intervalSec <= 30 {
+		num = 30
+	} else if intervalSec <= 60 {
+		num = 1
+		timeUnit = segutils.TMMinute
+	} else if intervalSec <= 300 {
+		num = 5
+		timeUnit = segutils.TMMinute
+	} else if intervalSec <= 600 {
+		num = 10
+		timeUnit = segutils.TMMinute
+	} else if intervalSec <= 1800 {
+		num = 30
+		timeUnit = segutils.TMMinute
+	} else if intervalSec <= 3600 {
+		num = 1
+		timeUnit = segutils.TMHour
+	} else if intervalSec <= 86400 {
+		num = 1
+		timeUnit = segutils.TMDay
+	} else {
+		// maximum granularity is 1 month as per experiments
+		num = 1
+		timeUnit = segutils.TMMonth
+	}
+
+	estimatedSpan := &structs.BinSpanOptions{
+		BinSpanLength: &structs.BinSpanLength{
+			Num:       num,
+			TimeScale: timeUnit,
+		},
+	}
+
+	return estimatedSpan
+}
+
+// Initial method to perform bin request
+func performBinRequest(nodeResult *structs.NodeResult, letColReq *structs.LetColumnsRequest, recs map[string]map[string]interface{}, finalCols map[string]bool, recordIndexInFinal map[string]int, numTotalSegments uint64, finishesSegment bool) error {
+	if recs != nil {
+		if letColReq.BinRequest.BinSpanOptions != nil {
+			return performBinRequestOnRawRecordWithSpan(nodeResult, letColReq, recs, finalCols)
+		} else {
+			return performBinRequestOnRawRecordWithoutSpan(nodeResult, letColReq, recs, finalCols, recordIndexInFinal, numTotalSegments, finishesSegment)
+		}
+	}
+
+	if len(nodeResult.Histogram) > 0 {
+		err := performBinRequestOnHistogram(nodeResult, letColReq)
+		if err != nil {
+			return fmt.Errorf("performBinRequest: Error while performing bin request on histogram, err: %v", err)
+		}
+	}
+
+	if len(nodeResult.MeasureResults) > 0 {
+		err := performBinRequestOnMeasureResults(nodeResult, letColReq)
+		if err != nil {
+			return fmt.Errorf("performBinRequest: Error while performing bin request on measure results, err: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func performBin(value float64, spanOptions *structs.BinSpanOptions, binReq *structs.BinCmdOptions) (interface{}, error) {
```

Comment:
May be refactor this to `performBinWithSpanOptions`. Since you are checking specifically for `SpanOptions`.

Reply:
Wouldn't this name be confusing, because we already have performBinWithSpan and perfrormBinWithSpanTime?

Reply:
addressed in the new commit.